src/tallies/filter_azimuthal.cpp

```diff
@@ -0,0 +1,85 @@
+#include "openmc/tallies/filter_azimuthal.h"
+
+#include <cmath>
+#include <sstream>
+
+#include "openmc/constants.h"
+#include "openmc/error.h"
+#include "openmc/search.h"
+#include "openmc/xml_interface.h"
+
+namespace openmc {
+
+void
+AzimuthalFilter::from_xml(pugi::xml_node node)
+{
+  auto bins = get_node_array<double>(node, "bins");
+
+  if (bins.size() == 1) {
+    // Allow a user to input a lone number which will mean that you subdivide
+    // [-pi,pi) evenly with the input being the number of bins
+
+    int n_angle = bins[0];
+    if (n_angle <= 1) throw std::runtime_error{
+      "Number of bins for azimuthal filter must be greater than 1."};
+
+    double d_angle = 2.0 * PI / n_angle;
+    bins.resize(n_angle + 1);
+    for (int i = 0; i < n_angle; i++) bins[i] = -PI + i * d_angle;
+    bins[n_angle] = PI;
+  }
+
+  this->set_bins(bins);
+}
+
+void AzimuthalFilter::set_bins(gsl::span<double> bins)
+{
+  // Clear existing bins
+  bins_.clear();
+  bins_.reserve(bins.size());
+
+  // Copy bins, ensuring they are valid
+  for (gsl::index i = 0; i < bins.size(); ++i) {
+    if (i > 0 && bins[i] <= bins[i-1]) {
+      throw std::runtime_error{"Azimuthal bins must be monotonically increasing."};
+    }
+    bins_.push_back(bins[i]);
+  }
+
+  n_bins_ = bins_.size() - 1;
+}
+
+void
+AzimuthalFilter::get_all_bins(const Particle* p, int estimator,
+                              FilterMatch& match) const
+{
+  double phi;
+  if (estimator == ESTIMATOR_TRACKLENGTH) {
```

Comment:
If x and y are zero, this can throw an error for std::atan2. We probably want to catch and handle this error.